crates/types/src/lib.rs

```diff
@@ -40,6 +37,110 @@ pub use receipt::{
 /// `Addressable` types
 pub use address::{Address, TemplateAddr};
 
+/// Operations on fixed-size byte array entities.
+pub trait BytesPrimitive<const N: usize>:
+    From<[u8; N]> + AsRef<[u8; N]> + std::hash::Hash + PartialEq + Eq + Clone
+{
+    /// Copies the contents of `val` in a new instance of `Self`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `val.as_ref()` is not of the appropriate length.
+    fn new<V>(val: V) -> Self
+    where
+        V: AsRef<[u8]>,
+    {
+        Self::new_opt(val).expect("Invalid bytes length")
+    }
+
+    /// Copies the contents of `val` in a new instance of `Self`; returns None
+    /// in case `val.as_ref()` is not of the appropriate length.
+    fn new_opt<V>(val: V) -> Option<Self>
+    where
+        V: AsRef<[u8]>,
+    {
+        let bytes = val.as_ref().try_into().ok()?;
+        Some(Self::from(bytes))
+    }
+
+    /// Creates a new `Self` that starts with the UTF-8 byte representation of
+    /// `s`; all remaining bytes are set to zero.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `s` doesn't fit into `Self`.
+    fn of(s: &str) -> Self {
+        let bytes = s.as_bytes();
+        assert!(bytes.len() <= N);
+
+        let mut buf = [0u8; N];
+        (&mut buf[..bytes.len()]).clone_from_slice(bytes);
+
+        Self::from(buf)
+    }
+
+    /// Returns a reference to the raw contents of `self`.
+    fn as_slice(&self) -> &[u8] {
+        self.as_ref()
+    }
+
+    /// Creates a new `Self` with all bytes set to zero.
+    fn zeros() -> Self {
```

Comment:
can we turn it into `const fn` ?

Reply:
We will in the future, for now Rust doesn't support `const fn` in traits.